src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,248 @@
+        }
+        if (hypers->alpha0 <= 0) {
+            throw std::invalid_argument("Scale parameter must be > 0");
+        }
+        if (phi->beta <= 0) {
+            throw std::invalid_argument("Diffusion parameter must be > 0");
+        }
+        if (phi->q <= 0) {
+            throw std::invalid_argument("Number of factors must be > 0");
+        }
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
+void MFAHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState>& states) {
+    auto& rng = bayesmix::Rng::Instance().get();
+    if (prior->has_fixed_values()) {
+        return;
+    }
+
+    else {
+        throw std::invalid_argument("Unrecognized hierarchy prior");
+    }
+}
+
+void MFAHierarchy::update_summary_statistics(const Eigen::RowVectorXd& datum,
+    bool add) {
+    if (add) {
+        data_sum += datum;
+        data_sum_squares += datum .cwiseProduct(datum);
+        data.push_back(datum);
+    }
+    else {
+        data_sum -= datum;
+        data_sum_squares -= datum.cwiseProduct(datum);
+        auto pos = std::find(data.begin(), data.end(), datum)
+        data.erase(pos);
+    }
+}
+
+void MFAHierarchy::clear_summary_statistics() {
+    data_sum = Eigen::VectorXd::Zero(p);
+    data_sum_squares = Eigen::VectorXd::Zero(p);
+    data.clear();
+}
+
+void MFAHierarchy::set_state_from_proto(
+    const google::protobuf::Message& state_) {
+    auto& statecast = downcast_state(state_);
+    state.mu = statecast.mfa_state().mu(); // TODO add, ls??
+    state.psi = statecast.mfa_state().psi();
+    state.eta = statecast.mfa_state().eta();
+    state.lambda = statecast.mfa_state().lambda();
+    set_card(statecast.cardinality());
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+MFAHierarchy::get_state_proto() const {
+    bayesmix::MFAState state_;
+    state_.set_mu(state.mu);
+    state_.set_psi(state.psi);
+    state_.set_Eta(state.Eta);
+    state_.set_Lambda(state.Lambda);
+
+    auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+    out->mutable_mfa_state()->CopyFrom(state_);
+    return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+    auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+    hypers->mutilde = hyperscast.mutilde();
+    hypers->alpha0 = hyperscast.alpha0();
+    hypers->beta = hyperscast.beta();
+    hypers->phi = hyperscast.phi();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+    bayesmix::MFAState hypers_;
+    hypers_.set_mutilde(hypers->mutilde);
+    hypers_.set_alpha0(hypers->alpha0);
+    hypers_.set_beta(hypers->beta);
+    hypers_.set_phi(hypers->phi);
+
+    auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+    out->mutable_mfa_state()->CopyFrom(hypers_);
+    return out;
+}
+
```

Comment:
What is the meaning of update_params in a non-conjugate hierarchy? We interpreted it as if update_params=true meant that when we sample from full conditionals, updated available parameters are always used for new parameters, for example, if we sample from Eta and then we sample from mu the new values of Eta are going to be used, while if update_params=false we always use the current state parameters, is this correct?

Reply:
That's a very good point, it does not make sense! In particular, it will always be false. So you could just add a check (with an assert) and raise an error if  `update_params==true`. `sample_full_cond` should update the current state (saved in the variable `State state`), either whichever algorithm you want. Since the paper uses a Gibbs Sampler to sample from Eta, then mu, then psi, then Lambda, this is exactly what you should implement.
However, you should not create a new empty state if `update_params` is false. So there should be only one state and you keep updating it one piece (i.e., parameter) at a time, conditioning on the current values of all the other parameters. Specific comments follow